packages/block-library/src/table/block.json

```diff
@@ -127,7 +127,27 @@
 		"align": true,
 		"color": {
 			"__experimentalSkipSerialization": true,
-			"gradients": true
+			"gradients": true,
+			"sets": [
```

Comment:
Hey, I appreciate the effort to prepare the proposal.

I think I've mentioned this in other places but can't find now the link. The most promising path to add _more_ style options to a block is exploring the elements API. Right now, it works in `theme.json` and only with a limited set (link and headers). We could look into expanding this to more elements and port it to block supports. This idea is in line with [design explorations for the global styles interface](https://github.com/WordPress/gutenberg/issues/27473#issuecomment-796773866). I presume elements should also have an equivalent in the block styles interface.

We've explored in the past the idea of adding [translations](
https://github.com/WordPress/gutenberg/pull/29155#issuecomment-802845248) (e.g.: more metadata) to block supports but the overall thinking was that it should remain a lean API.

Reply:
Actually. It'd be good to have some confirmation about direction here, in case I've missed some context on this. cc @jasmussen @mtias how do you see us enabling _more_ style choices for blocks? Are elements the way to go?

Reply:
Can you expand a bit what you mean by "more style choices"? Do you mean like how #31149 has both background color and overlay background color on the same block?

Reply:
Thanks for the direction @nosolosw 🙇 

> The most promising path to add more style options to a block is exploring the elements API. Right now, it works in theme.json and only with a limited set (link and headers).

I've misunderstood the intent of the elements API then. 

Would you be happy for it to contain CSS selectors that are more complex than simple DOM elements such as `h1`, `h2`, `h3` etc? The elements API might also need updating so CSS pseudo classes such as `:hover` could be used. That would allow a secondary set of colors to be applied to a button for example.

> We've explored in the past the idea of adding translations (e.g.: more metadata) to block supports but the overall thinking was that it should remain a lean API.

Ok, I found https://github.com/WordPress/gutenberg/pull/30293 which made it look like allowing the translation of the secondary color labels would just be a matter of adding an entry into `packages/blocks/src/api/i18n-block.json` to cover the label fields.

How would we allow blocks to define a label for a secondary set of colors, if using the elements API? For example, header, footer and striped colors for tables, active/hover colors for buttons etc.

Reply:
> Can you expand a bit what you mean by "more style choices"?

Yeah, I may have been too concise. Let me expand how I think about the next steps.

A block can always add a custom panel to its block sidebar as in https://github.com/WordPress/gutenberg/pull/31149 That's fine. We need to allow blocks to register custom panels in the global styles sidebar as well: this is a feature we don't have at the moment and that may need to wait a bit until the UI for global styles is more developed (some progress at https://github.com/WordPress/gutenberg/pull/32923).

The alternative approach blocks have, and what I meant by "more style choices", is the block supports API. What are the next steps for it? This PR proposes to expand this API by adding the concept of "sets". My question is: shouldn't we develop the already existing concept of "elements" instead? As I see it, there're two things to do as a follow-up to https://github.com/WordPress/gutenberg/issues/29891:

- UI for elements: elements have a matching UI panel for them. It's already mocked up for the [global styles sidebar](https://github.com/WordPress/gutenberg/issues/27473#issuecomment-796773866) but it's not for the block sidebar. Shouldn't the users be able to style elements a block supports via the block sidebar as well?

- The HTML representation of elements: while elements have a default serialization to HTML (`link` is serialized as the `a` CSS selector), blocks should be able to hook into this to modify it.

By doing these two things, we enable users to interact with the global styles sidebar in a more semantic way: when users update the link element in the global styles sidebar they are not changing the `a` elements but they are styling links across the site - no matter how they are serialized to HTML. At the same time, users will also gain the ability to change elements that are part of complex blocks, such as table, search, navigation, etc.

As/If we walk down this path, we also need to look at what other elements are necessary: perhaps header and footer elements? Are there blocks that would benefit from these other than the table block? etc.

That's how I see things evolving. How do others feel about this?

Reply:
At a practical level, if we go the elements route, I presume it'd take some time to mature. In the meantime, I'd default to adding custom controls for blocks when the block supports API isn't enough. We can always refactor later to the new APIs we introduce.

Reply:
Thanks for expanding on this @nosolosw.


> Shouldn’t the users be able to style elements a block supports via the block sidebar as well?

It makes sense that you should be able to via the block sidebar.

> The HTML representation of elements: while elements have a default serialization to HTML (link is serialized as the a CSS selector), blocks should be able to hook into this to modify it.

This might handle mapping elements to appropriate CSS selectors. How do we get these translated when they relate to something a little more complicated such as a button’s hover state or the nth-child row of a table?

Another question is if elements are to be a closed set how do we scale this to allow different blocks to introduce their own requirements?  

The table today might want header, footer and striped colors. Later, it might be desired to have a different set of colors for the row when it is hovered. Those then need different elements to perhaps ones that could be added for a nav block, maybe active/hover states, dropdown menu backgrounds etc.

I think there could be a third step required, allowing configuration and translation of labels for those element controls.

> In the meantime, I’d default to adding custom controls for blocks when the block supports API isn’t enough.

It was taking such an approach that lead to this PR. 

A simplistic PR [adding ad hoc colors to the table](https://github.com/WordPress/gutenberg/pull/31696) ( [issue](https://github.com/WordPress/gutenberg/issues/31261) ) was created to generate further discussion on how we could achieve secondary color configuration and application in a standard way. Not just for the table but any block. 

The result of adding the desired controls to the table block in this case is a monster panel that doesn't have much chance of landing. It also then ends up with two panels dedicated to controlling color. 

It would be good to find some middle ground where a temporary solution could be implemented before switching to the new APIs as suggested.

Reply:
This is an issue that affects more blocks, such as [navigation](https://github.com/WordPress/gutenberg/pull/31149#issuecomment-877079245). I shared with some folks (Riad, Matías) and they agree we should avoid extending our block supports or elements to add custom things at the moment. Instead, we should resort to custom controls if we need them to make progress.

I know that approach comes with the issue that these custom controls/styles can't be targeted via the global styles sidebar (users) and `theme.json` (theme authors). This is something we need to fix. As a starting point, I've gathered my thoughts at https://github.com/WordPress/gutenberg/issues/33255 so we have a place to discuss it.

That's the overall stance for this. However, for this specific case, have you thought about converting the table to inner blocks? I know that approach can result in a suboptimal UX in some blocks that have many similar children, but I wonder how would it fare here.

Reply:
I wanted to note that I believe https://github.com/WordPress/gutenberg/pull/32659 is a key ingredient in allowing global styles to style more properties. Right now one challenge is high specificity of default block styles, which is mitigated by that PR.

Reply:
Thanks for sharing, I'll read that one (no need to wait for me to merge, though, I see it's already approved).